core/src/main/java/org/infinispan/persistence/spi/NonBlockingStore.java

```diff
@@ -0,0 +1,379 @@
+    * {@link #entryPublisher(IntSet, Predicate)} is already implemented.
+    * <pre>{@code
+    * return Flowable.fromPublisher(entryPublisher(segments, filter))
+    *     .map(MarshallableEntry::getKey);}
+    * </pre>
+    * @param segments A set of segments to filter keys by. If a null is provided, consider it as all segments.
+    * @param filter A filter to filter they keys by. If this is null then no additional filtering should be done after segments.
+    * @return a publisher that will provide the keys from the store
+    */
+   default Publisher<K> publishKeys(IntSet segments, Predicate<? super K> filter) {
+      throw new UnsupportedOperationException("Store characteristic included " + Characteristic.BULK_READ + ", but it does not implement entryPublisher");
+
+   }
+
+   /**
+    *
+    * @return
+    */
+   default Publisher<MarshallableEntry<K, V>> purgeExpired() {
+      throw new UnsupportedOperationException("Store characteristic included " + Characteristic.EXPIRATION + ", but it does not implement purgeExpired");
+   }
+
+   /**
+    * Write modifications to the store in the prepare phase, as this is the only way we know the FINAL values of the entries.
+    * This is required to handle scenarios where an objects value is changed after the put command has been executed, but
+    * before the commit is called on the Tx.
+    *
+    * @param transaction the current transactional context.
+    * @param batchModification an object containing the write/remove operations required for this transaction.
+    */
+   default CompletionStage<Void> prepareWithModifications(Transaction transaction, BatchModification batchModification) {
+      throw new UnsupportedOperationException("Store characteristic included " + Characteristic.TRANSACTIONAL + ", but it does not implement rollback");
+   }
+
+   /**
+    * Commit the provided transaction's changes to the underlying store.
+    *
+    * @param transaction the current transactional context.
+    */
+   default CompletionStage<Void> commit(Transaction transaction) {
+      throw new UnsupportedOperationException("Store characteristic included " + Characteristic.TRANSACTIONAL + ", but it does not implement rollback");
+   }
+
+   /**
+    * Rollback the provided transaction's changes to the underlying store.
+    *
+    * @param transaction the current transactional context.
+    */
+   default CompletionStage<Void> rollback(Transaction transaction) {
+      throw new UnsupportedOperationException("Store characteristic included " + Characteristic.TRANSACTIONAL + ", but it does not implement rollback");
+   }
+
+   /**
+    * Some stores may not want to perform a write operation if a command has certain flags
+    * @implSpec
+    * The default implementation returns true
+    * @param commandFlags the flags attributed to the command when performing the write
+    * @return whether the write should occur
+    */
+   default boolean shouldWrite(long commandFlags) {
```

Comment:
use case? 
use a `Set<Flag>` instead to make it easier for users?

Reply:
The only use case currently is https://github.com/infinispan/infinispan/blob/master/persistence/remote/src/main/java/org/infinispan/persistence/remote/RemoteStore.java#L454

Unfortunately changing it to Set<Flag> probably isn't going to be a great idea as we internally store the flags as this long value on the command, which means every write would be creating an EnumSet from the long.

I think I will leave this for now, but put it in the back of my mind how we can make this cleaner.

Reply:
oh... in that case, it would be better to have a "RollingUpgradeStore" (read-only).
internally, it could use Hot Rod or Rest

Reply:
Sorry, I am not quite sure what you mean here. Are you saying we should have a new concrete implementation for rolling upgrades? And how would the calling flags be passed still?

Reply:
my idea was to have `RollingUpgradeStore` as shared, read-only store (with "bulk-read" support). In my head, we would never need to write to this store and `shouldWrite(long commandFlags)` could be removed. I'm wrong!
just ignore me.

Reply:
No worries, I was hoping we could figure out something better.

Reply:
Instead of `shouldWrite` how about `boolean ignoreCommandWithFlags`? I think it makes the commands purpose more explicit.

Reply:
Hrmm, sure. I am not too picky about this command as I doubt any users will actually implement it.